addons/mass_mailing/models/mass_mailing.py

```diff
@@ -555,14 +604,22 @@ def copy(self, default=None):
     def _group_expand_states(self, states, domain, order):
         return [key for key, val in type(self).state.selection]
 
-    def update_opt_out(self, email, res_ids, value):
-        model = self.env[self.mailing_model_real].with_context(active_test=False)
-        if 'opt_out' in model._fields:
-            email_fname = 'email_from'
-            if 'email' in model._fields:
-                email_fname = 'email'
-            records = model.search([('id', 'in', res_ids), (email_fname, 'ilike', email)])
-            records.write({'opt_out': value})
+    def update_opt_out(self, email, list_ids, value):
+        if len(list_ids) > 0:
+            model = self.env['mail.mass_mailing.contact'].with_context(active_test=False)
+            records = model.search([('email', '=ilike', email)])
+            opt_out_records = self.env['mail.mass_mailing.list_contact_rel']\
+                .search([('contact_id', 'in', records.ids)])\
+                .filtered(lambda rec: rec.list_id.id in list_ids and rec.opt_out != value)
+
+            opt_out_records.write({'opt_out': value})
+            message = 'The recipient <strong>unsubscribed from %s</strong> mailing list(s)' \
+                if value else 'The recipient <strong>subscribed to %s</strong> mailing list(s)'
+            for record in records:
+                # filter the list_id by record
+                record_lists = opt_out_records.filtered(lambda rec: rec.contact_id.id == record.id)
```

Comment:
Whut ? It seems this code can be simplified a bit, record, records, record_lists, ... just have the right record set at the beginning of the computation and update it :) .

Reply:
this objective is to log correctly on which mailing list the mass_mailing.contact have been opted_out or in. If 2 mass_mailing contact with same email are subscribed to different mailing list, we should not log on contact A that he has been opted_out on mailing list from contact B.